logsapi/functionlogs.go

```diff
@@ -0,0 +1,118 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package logsapi
+
+import (
+	"github.com/elastic/apm-aws-lambda/apmproxy"
+	"go.elastic.co/apm/v2/model"
+	"go.elastic.co/fastjson"
+)
+
+type logContainer struct {
+	Log *logLine
+}
+
+type logLine struct {
+	Timestamp model.Time
+	Message   string
+	FAAS      *faas
+}
+
+func (l *logLine) MarshalFastJSON(w *fastjson.Writer) error {
+	var firstErr error
+	w.RawString("{\"message\":")
+	w.String(l.Message)
+	w.RawString(",\"@timestamp\":")
+	if err := l.Timestamp.MarshalFastJSON(w); err != nil && firstErr == nil {
+		firstErr = err
+	}
+	if l.FAAS != nil {
+		w.RawString(",\"faas\":")
+		if err := l.FAAS.MarshalFastJSON(w); err != nil && firstErr == nil {
+			firstErr = err
+		}
+	}
+	w.RawByte('}')
+	return firstErr
+}
+
+// faas struct is a subset of go.elastic.co/apm/v2/model#FAAS
+//
+// The purpose of having a separate struct is to have a custom
+// marshalling logic that is targeted for the faas fields
+// available for function logs. For example: `coldstart` value
+// cannot be inferred for function logs so this struct drops
+// the field entirely.
```

Comment:
Doesn't the extension get initialised once per cold start? Meaning that we could set `coldstart: true` for the first invocation processed by the extension, and `coldstart: false` for every request thereafter?

We don't need to change this now, but I think we _could_ improve it in the future. We probably don't need to set this property on every log message anyway, though.

Reply:
> Doesn't the extension get initialised once per cold start? Meaning that we could set coldstart: true for the first invocation processed by the extension, and coldstart: false for every request thereafter?

Hmm, that's an amazing idea. I will create an issue for this.